chain-signatures/node/src/protocol/posit.rs

```diff
@@ -0,0 +1,270 @@
+use cait_sith::protocol::Participant;
+use lru::LruCache;
+use serde::{Deserialize, Serialize};
+
+use std::collections::HashSet;
+use std::fmt;
+use std::hash::Hash;
+use std::num::NonZeroUsize;
+
+pub type ProposerId = Participant;
+
+pub enum Positor<T> {
+    Proposer(ProposerId, T),
+    Deliberator(ProposerId),
+}
+
+impl<T> Positor<T> {
+    pub fn is_proposer(&self) -> bool {
+        matches!(self, Positor::Proposer(_, _))
+    }
+
+    pub fn id(&self) -> ProposerId {
+        match self {
+            Positor::Proposer(id, _) => *id,
+            Positor::Deliberator(id) => *id,
+        }
+    }
+}
+
+/// All actions that can be taken when a new posit is introduced for a protocol.
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+pub enum PositAction {
+    Propose,
+    Start(Vec<Participant>),
+    Accept,
+    // TODO: Reject can also have a reason
+    Reject,
+}
+
+impl PositAction {
+    pub fn is_accept(&self) -> bool {
+        matches!(self, PositAction::Accept)
+    }
+}
+
+pub enum PositInternalAction<S> {
+    StartProtocol(Vec<Participant>, Positor<S>),
+    Reply(PositAction),
+    None,
+}
+
+/// A counter for a posit. This is used to track the participants that have
+/// accepted the posit alongside storing an intermediary state for the protocol
+/// that the proposer needs to keep track of.
+pub struct PositCounter<S> {
+    pub participants: HashSet<Participant>,
+    accepts: HashSet<Participant>,
+    rejects: HashSet<Participant>,
+    store: S,
+}
+
+/// A collection of posits that are being proposed. This is used to track
+/// the posits that are being proposed and the participants that have
+/// accepted them.
+pub struct Posits<Id, S> {
+    me: Participant,
+
+    // TODO: probably should be a LruCache to expire items
+    /// The posits that our node has proposed.
+    posits: LruCache<Id, Positor<PositCounter<S>>>,
+}
+
+impl<T: Copy + Hash + Eq + fmt::Debug, S> Posits<T, S> {
+    pub fn new(me: Participant) -> Self {
+        Self {
+            me,
+            // 1024 is a good default size for the cache since we don't expect to be have more
+            // than 1024 concurring posits at any time. This should however be configurable in
+            // the future.
+            posits: LruCache::new(NonZeroUsize::new(1024).unwrap()),
```

Comment:
Can you explain a bit more about the motivation / benefits for the LRU cache?  (instead of an unlimited set like HashSet)

I guess strictly limited memory usage can be nice. But shouldn't we be deliberate about which posits we abandon?

I think dropping posits as it is coded now can hurt robustness in certain fault scenarios.
For example, a malfunctioning node that sends many `PositAction::Propose` messages will flush out valid posits on all its peers. This can potentially halt the entire system, if all other nodes can't get posits accepted before they are flushed out of the LRU cache.

And if dropping an entry from self.posits is equivalent to going from `Prepare` back to `init`, this also conflicts with the proof table in the specification.

Reply:
that's true, it's definitely better to have our own explicit cleanup logic here too. I'll just change it back to hashmap. Initially just had the lru cache just so there weren't many leftover posits that never get cleaned up an "leak" memory but the storage for them is decently small (a `u32` per entry) anyways.